napari/utils/colormaps/colormap_utils.py

```diff
@@ -618,7 +618,30 @@ def ensure_colormap(colormap: ValidColormapArg) -> Colormap:
     """
     with AVAILABLE_COLORMAPS_LOCK:
         if isinstance(colormap, str):
-            name = colormap
+            # is a colormap with this name (default or custom) already available?
+            cmap = AVAILABLE_COLORMAPS.get(
+                colormap,
+                AVAILABLE_COLORMAPS.get(f"custom-{colormap.lower()}", None),
+            )
+            if cmap is None:
+                cmap = _colormap_from_colors(
+                    colormap, name=f"custom-{colormap.lower()}"
+                )
+                if cmap is None:
+                    cmap = vispy_or_mpl_colormap(colormap)
+
+                for cmap_ in AVAILABLE_COLORMAPS.values():
+                    if (
+                        np.array_equal(cmap_.controls, cmap.controls)
```

Comment:
This line requires fixing type annotation of `napari.utils.colormaps.colormap.Colormap` (field `controls`). As the validators of this field, ensure that this field is always Array, but type annotation state that it is `Optional[Array]`

Reply:
This one I have trouble understanding, sorry.

Let me know if this is acceptable: I've modified my test starting at line 633 to check whether either of `cmap_.controls` or `cmap.controls` (actually renamed `custom_cmap` as per explanation above) is `None` before my `np.equal` condition.

```python
                for cmap_ in AVAILABLE_COLORMAPS.values():
                    if (
                        cmap_.controls is not None
                        and custom_cmap.controls is not None
                        and np.array_equal(cmap_.controls, custom_cmap.controls)
                        and cmap_.colors.shape == custom_cmap.colors.shape
                        and np.all(cmap_.colors == custom_cmap.colors)
                        and cmap_.interpolation == custom_cmap.interpolation
                    ):
                        custom_cmap = cmap_
                        break
```

Reply:
change line 54 of `napari/utils/colormaps/colormap.py` to:

```python
controls: Array = Field(default_factory=lambda :Array([]))
```